packages/ensnode-react/src/utils/query.ts

```diff
@@ -0,0 +1,87 @@
+"use client";
+
+import {
+  ENSNodeClient,
+  type ForwardResolutionSelection,
+  type ForwardResponse,
+  type ReverseResponse,
+} from "@ensnode/ensnode-sdk";
+import { useQuery } from "@tanstack/react-query";
+import type { ENSNodeConfig, UseQueryReturnType } from "../types";
+
+/**
+ * Query key factory for ENSNode queries
+ * Includes endpoint URL for proper cache isolation between different ENSNode instances
+ *
+ * @example
+ * ```typescript
+ * // Forward resolution
+ * queryKeys.forward("https://api.mainnet.ensnode.io", "vitalik.eth", { addresses: [60] })
+ * // Results in: ["ensnode", "https://api.mainnet.ensnode.io", "resolution", "forward", "vitalik.eth", { addresses: [60] }]
+ *
+ * // Reverse resolution
+ * queryKeys.reverse("https://api.mainnet.ensnode.io", "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045", 1)
+ * // Results in: ["ensnode", "https://api.mainnet.ensnode.io", "resolution", "reverse", "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045", 1]
+ *
+ * // Different endpoints have separate cache keys
+ * queryKeys.forward("https://api.testnet.ensnode.io", "vitalik.eth", { addresses: [60] })
+ * // Results in: ["ensnode", "https://api.testnet.ensnode.io", "resolution", "forward", "vitalik.eth", { addresses: [60] }]
+ * ```
+ */
+export const queryKeys = {
+  all: (endpointUrl: string) => ["ensnode", endpointUrl] as const,
+  resolutions: (endpointUrl: string) => [...queryKeys.all(endpointUrl), "resolution"] as const,
+  forward: (endpointUrl: string, name: string, selection?: ForwardResolutionSelection) =>
+    [...queryKeys.resolutions(endpointUrl), "forward", name, selection] as const,
+  reverse: (endpointUrl: string, address: string, chainId?: number) =>
+    [...queryKeys.resolutions(endpointUrl), "reverse", address, chainId] as const,
+};
+
+/**
+ * Create query options for forward resolution (name to records)
+ */
+export function createForwardResolutionQueryOptions(
+  config: ENSNodeConfig,
+  name: string,
+  selection?: ForwardResolutionSelection,
+) {
+  return {
+    queryKey: queryKeys.forward(config.client.endpointUrl.href, name, selection),
+    queryFn: async (): Promise<ForwardResponse> => {
+      const client = new ENSNodeClient(config.client);
+      return client.resolveName(name, selection);
+    },
+    enabled: Boolean(name),
+    staleTime: 1000 * 60 * 5, // 5 minutes
```

Comment:
we should allow the users to control the caching behavior on a per-hook level as part of some common options, right? especially to disable cache on a per-hook basis for apps that are directly affecting the ENS protocol, they might expect changes on a faster timeframe than 5m

also: move to documented shared constant (can stay in this file)

Reply:
likewise for `gcTime`

Reply:
ah perhaps instead the correct pattern is for those users to use `refetch` in order to get freshest data? is there a good argument for per-hook staleTime/gcTime configurability? should we just allow users to pass arbitrary common react query options to our hooks along with our hook-specific arguments and merge them in here?